controller/state/statedb.go

```diff
@@ -511,26 +519,33 @@ func (s *stateDB) NewRetrievalTask(ctx context.Context, retrievalTask tasks.Retr
 }
 
 func (s *stateDB) TaskHistory(ctx context.Context, taskID string) ([]tasks.TaskEvent, error) {
-	rows, err := s.db().QueryContext(ctx, taskHistorySQL, taskID)
-	if err != nil {
-		return nil, err
-	}
-	defer rows.Close()
-
 	var history []tasks.TaskEvent
-	for rows.Next() {
-		var status, run int
-		var ts time.Time
-		var stage string
-		if err = rows.Scan(&status, &stage, &run, &ts); err != nil {
-			return nil, err
+	err := s.transact(ctx, func(tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, taskHistorySQL, taskID)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+
+		history = nil
```

Comment:
Non-blocking:
I assume this line is cause `transact` may retry -- we should probably indicate somewhere that retries are possible and if you're using closures to access higher scoped vars you better reset them. took me a few rereads to figure out what this mean or why it was needed.

Reply:
Nice deduction!  That's exactly why.  I will add comments.

Reply:
Done